Homework DB week 2 Vladimir

intruston opened the pull request.

Breus commented on Week2/exercises/1createTableAuthors.js:
```diff
@@ -0,0 +1,30 @@
+    if (err) throw err;
+    console.log("Connected to the database");});
+
+const sql = `CREATE TABLE authors (
+  author_id INT AUTO_INCREMENT PRIMARY KEY,
+  author_name VARCHAR(255) NOT NULL,
+  university VARCHAR(255),
+  date_of_birth DATE,
+  h_index INT,
+  gender ENUM('male', 'female', 'other')
```
Correct!

Breus commented on Week2/exercises/1createTableAuthors.js:
```diff
@@ -0,0 +1,30 @@
+  gender ENUM('male', 'female', 'other')
+)`;
+
+connection.query(sql, function (error, results) {
+  if (error) throw error;
+  console.log('Table created successfully!');
+});
```
Quite a lot of duplicated code with `connection.query, if/else, etc.`. Can you think of a way to get less duplicated code?

intruston replied:
Added connection.js file to fix that!

Breus commented on Week2/exercises/3addMentorAuthors.js:
```diff
@@ -0,0 +1,30 @@
+const mysql = require('mysql');
+
+const connection = mysql.createConnection({
```
This is actually question 2 of exercise 1 (keys). Could you maybe order your files like the questions? Now the file has number 3, but that doesn't correspond to the assignment description.

intruston replied:
Filenames are updated

Breus commented on Week2/exercises/3addMentorAuthors.js:
```diff
@@ -0,0 +1,30 @@
+    if (err) throw err;
+    console.log("Connected to the database");});
+
+const sql = `ALTER TABLE authors ADD mentor INT`;
+
+connection.query(sql, function (error, results) {
+  if (error) throw error;
+  console.log('Add mentor');
+});
+
+const sql2 = `ALTER TABLE authors ADD FOREIGN KEY (mentor) REFERENCES authors(author_id);`;
```
Correct altering of the authors table, but you can add the mentor column and the foreign key constraint in one query.

intruston replied:
Writed 1 alter table for that.

Breus commented on Week2/exercises/2createTableresearch_papers.js:
```diff
@@ -0,0 +1,29 @@
+    paper_title VARCHAR(255) NOT NULL,
+    conference VARCHAR(255),
+    publish_date DATE,
+    author_id INT
```
The research papers table should not contain an `author_id` column: this would assume that every paper is written by just one author. However, in reality, multiple authors can write the same paper.

Therefore, paper/author relationship is a many-to-many relationship. Please update your table structure accordingly or optionally add a new table.

intruston replied:
Added author_paper table to add many-to-many relationships

Breus commented on Week2/exercises/4addForeignKeyreseach_papers.js:
```diff
@@ -0,0 +1,23 @@
+    if (err) throw err;
+    console.log("Connected to the database");});
+
+const sql = `ALTER TABLE research_papers ADD FOREIGN KEY (author_id) REFERENCES authors(author_id);`;
```
As said before: the research papers table shouldn't directly contain the author_id as this would be a 1-to-1 relationship between a paper and an author.

intruston replied:
Added author_paper table to add many-to-many relationship

Breus commented on Week2/exercises/5fillTables.js:
```diff
@@ -0,0 +1,63 @@
+];
+  
+  
+authors.forEach(author => {
+    connection.query('INSERT INTO authors SET ?', author, function (error, results) {
+      if (error) throw error;
+      console.log(`Author ${author.author_name} added!`);
+    });
+  });
+  
+research_papers.forEach(paper => {
+    connection.query('INSERT INTO research_papers SET ?', paper, function (error, results) {
+      if (error) throw error;
+      console.log(`Research paper ${paper.paper_title} added!`);
+    });
+  });
```
Inserting like this works, but it would be nicer if you could read the data from separate JSON files instead of JavaScript arrays directly in the same file.

intruston replied:
Added data.json file for that!

Breus commented on Week2/exercises/7selectResults.js:
```diff
@@ -0,0 +1,58 @@
+    if (err) throw err;
+    console.log("Connected to the database");});
+
+// Write a query that prints names of all authors and their corresponding mentors.
+const printAuthorsAndMentors = () => {
+    const sql = `
+        SELECT authors.author_name, mentor.author_name as mentor_name
+        FROM authors
+        LEFT JOIN authors as mentor ON authors.mentor = mentor.author_id;
+        `;
+
+    connection.query(sql, function (error, results, fields) {
+        if (error) throw error;
+          console.log("Authors and their mentors:");
+          results.forEach((result) => {
+            console.log(`${result.author_name} - Mentor: ${result.mentor_name}`);
+          });
+        });
+};
+
+//Write a query that prints all columns of authors and their published paper_title. If there is an author without any research_Papers, print the information of that author too.
+const printAuthorsAndPapers = () => {
+    const sql = `
+      SELECT authors.*, research_Papers.paper_title
+      FROM authors
+      LEFT JOIN research_Papers ON authors.author_id = research_Papers.author_id;
+    `;
+  
+    connection.query(sql, function (error, results, fields) {
```
Please update your tables according to the feedback I gave (many-to-many relationship) and update these select statements accordingly.

intruston replied:
Added author_paper table to add many-to-many relationship

Breus commented on Week2/exercises/8selectAggregateResults.js:
```diff
@@ -0,0 +1,107 @@
+    if (err) throw err;
+    console.log("Connected to the database");});
+
+//All research papers and the number of authors that wrote that paper.
```
Currently it would be of no use for you if I would review these queries since your table structure itself isn't correct yet for a many-to-many relationship. Please update the tables and all of these select queries accordingly.

intruston replied:
Added author_paper table to add many-to-many relationship

Breus reviewed (changes requested):
Hi @intruston, nice effort on your homework for week 2! However, since you didn't properly implement the many-to-many relationship in your table structure, the rest of your solutions were also incorrect.

Please make sure to correct your solutions. Best of luck!

intruston commented:
Hi @Breus. I have updated many-to-many relationship by adding author_paper table. I have also updated filenames, so now my filenames order is like questions order.
File [connection.js](https://github.com/HackYourHomework/databases-class41/pull/17/files#diff-928468053040b4a11b209b141c8ed3e9b3e010a07c43f8de28f1d68e1857ee86) was created to make code less duplicated.
File [data.json](https://github.com/HackYourHomework/databases-class41/pull/17/files#diff-9c1484393e92b435043ca48d82e8719e8e8368379953aedd2bf2ad4b29ad105c) was created so we could read the data from separate JSON files instead of JavaScript arrays directly in the same file.

Breus commented:
@intruston I've approved your homework now!
